### PR TITLE
[HPOS] Add HPOS support for the reserved stock query

### LINKED ISSUE #36375: [HPOS] Port reserved stock query to support HPOS.
This issue scopes migrating queries defined in class `ReserveStock` to support HPOS data tables, specifically in function `get_query_for_reserved_stock` and `reserve_stock_for_product`.

### PR DESCRIPTION
**Pull request by coreymckrill**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Enables the ReserveStock class to detect when HPOS tables are enabled and in use, and modify the query in `get_query_for_reserved_stock()` accordingly. Without this change, reserved stock is not correctly counted when using HPOS.

Closes #36375

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

This is tricky to test because when most orders are created, the stock gets reserved and then released within the space of the same HTTP request, so it's hard to isolate whether the reserved stock query is working or not.

To do it, we can use a simple mu-plugin:

```php
<?php

function test_stock_reserved() {
	$product_id = 119;
	$product = wc_get_product( $product_id );
	$reserved = wc_get_held_stock_quantity( $product );

	wc_get_logger()->debug( 'Reserved stock: ' . print_r( $reserved, true ) );
}
add_action( 'woocommerce_checkout_order_created', 'test_stock_reserved', 99 );
```

You'll need to add this code to your test site and change `119` to the ID of the product you'll be using in your test orders.

First observe the behavior before you check out this PR branch:

1. Ensure the product you're going to use for testing has inventory management enabled and enough stock available to make a few orders with.
1. Ensure your test site either has HPOS turned off (**WooCommerce > Settings > Advanced > Features**), or enabled but with the posts table selected as the data store, and with table synchronization enabled.
3. Turn on the COD payment method for easier testing.
4. Go through the front end checkout process to create an order containing the product whose ID you're using in the code snippet above.
5. After the checkout is complete, go to the logs screen (**WooCommerce > Status > Logs**). The most recent log should show `DEBUG Reserved stock: 1`
6. Now turn on HPOS, choose the WooCommerce orders tables as the data store, and turn off synchronization.
7. Make another order.
8. Now on the logs screen, there should be another entry, `DEBUG Reserved stock: 0`

The `0` in the second log entry indicates that no reserved stock was found during the request. This is the problem.

Now check out this branch and go back through the steps. This time, both logs should be `DEBUG Reserved stock: 1`, meaning that reserved stock was found even when using HPOS.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/src/Checkout/Helpers/ReserveStock.php [hunk 1]
```diff
@@ -202,18 +213,28 @@ private function reserve_stock_for_product( $product_id, $stock_quantity, $order
 	 */
 	private function get_query_for_reserved_stock( $product_id, $exclude_order_id = 0 ) {
 		global $wpdb;
+
+		$join         = "$wpdb->posts posts ON stock_table.`order_id` = posts.ID";
+		$where_status = "posts.post_status IN ( 'wc-checkout-draft', 'wc-pending' )";
+		if ( $this->is_cot_in_use() ) {
```
Comment (barryhughes) [inline plugins/woocommerce/src/Checkout/Helpers/ReserveStock.php]: Not essential you change this, but if you were to import `Automattic\WooCommerce\Utilities\OrderUtil` you could do:

```suggestion
		if ( OrderUtil::custom_orders_table_usage_is_enabled() ) {
```

(and then delete the `is_cot_in_use()` method.)

**Review by barryhughes — DISMISSED** [review]
LGTM, works as described 🚀 

I haven't merged as I left a note about an optional change: feel free to go ahead and merge if you prefer not to make that adjustment, or if you *do* tweak things ping me again for fresh approval :+1:

File: plugins/woocommerce/src/Checkout/Helpers/ReserveStock.php (hunk 1 above)
Comment (coreymckrill) [reply plugins/woocommerce/src/Checkout/Helpers/ReserveStock.php]: I like that better, thanks! 5d59e534dab35d24ac01205fb633c07bcf07aadc

**Review by barryhughes — APPROVED** [review]
🚀

**Comment by kahmeel** [comment]
[#36535] - limit of quantity set only works with single product but when it comes to variable products issue still exists. Count set on each variable products is not fixed. I set 7 quantity but when I try to add, it added more than 7

**Comment by coreymckrill** [comment]
@kahmeel Thanks for the heads up. Could you [open a new bug report](https://github.com/woocommerce/woocommerce/issues/new?assignees=&labels=&template=1-bug-report.yml) for this, with step-by-step on how to replicate the issue?

**Pull request merged**